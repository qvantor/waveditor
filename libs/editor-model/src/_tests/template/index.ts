export * as BasicCase from './basic-case';
export * as HostTemplate from './host-template';
