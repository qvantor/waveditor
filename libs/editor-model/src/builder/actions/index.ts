export { mergeComponent } from './merge-component';
