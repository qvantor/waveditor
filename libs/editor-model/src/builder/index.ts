export {
  createBuilderContext,
  builderContextToSnapshot,
  createEmptySnapshot,
} from './services';
export { BuilderProvider } from './constants';
export { useAction, useBuilderContext } from './hooks';
