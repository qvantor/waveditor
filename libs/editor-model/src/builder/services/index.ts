export { createBuilderContext } from './create-builder-core';
export { builderContextToSnapshot } from './mappers';
export { createEmptySnapshot } from './create-empty-snapshot';
