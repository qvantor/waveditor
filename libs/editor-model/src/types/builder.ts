import { Subject } from 'rxjs';
import { UndoRedoModule } from '@waveditors/rxjs-react';
import { ElementsStore } from '../elements';
import { HoverStore, SelectedStore } from '../interaction';
import { RelationsStore } from '../relations';
import { TemplateConfigStore } from '../template-config';
import { VariablesStore } from '../variables';
import { EditorEvents, EditorCommands, UndoRedoEvents } from './';

export type BuilderCore = Readonly<{
  model: {
    elements: ElementsStore;
    relations: RelationsStore;
    config: TemplateConfigStore;
    variables: VariablesStore;
  };
  interaction: {
    hover: HoverStore;
    selected: SelectedStore;
  };
  editor: {
    events: Subject<EditorEvents>;
    commands: Subject<EditorCommands>;
  };
  module: {
    undoRedo: UndoRedoModule<UndoRedoEvents>;
  };
  subscribe: () => () => void;
}>;

export type BuilderContext = Omit<BuilderCore, 'subscribe'>;
