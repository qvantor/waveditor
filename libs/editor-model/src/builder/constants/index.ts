import { createContext } from 'react';
import { BuilderContext } from '../../types';

export const BuilderContextValue = createContext<BuilderContext | null>(null);
export const BuilderProvider = BuilderContextValue.Provider;
