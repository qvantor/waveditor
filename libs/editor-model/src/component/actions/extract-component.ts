import { ElementStore, isLayoutStore, getElementById } from '../../elements';
import { EditorSnapshot, BuilderContext } from '../../types';
import { getLayoutChildren } from '../../elements/layout';

const getElements = (
  rootElement: ElementStore,
  elements: Record<string, ElementStore>
): EditorSnapshot['elements'] => {
  const root = rootElement.getValue();
  const elementsRoot = { [root.id]: root };
  if (!isLayoutStore(rootElement)) return elementsRoot;
  return getLayoutChildren(elements)(rootElement.getValue()).reduce(
    (sum, element) => {
      const elementValue = element.getValue();
      return { ...sum, [elementValue.id]: elementValue };
    },
    elementsRoot
  );
};

// function to collect subtree from BuilderContext into new component tree
export const extractComponent =
  ({ model: { elements, config } }: BuilderContext) =>
  (rootElementId: string): EditorSnapshot => {
    const elementsValue = elements.getValue();
    const configValue = config.getValue();
    const rootElement = getElementById(rootElementId)(
      elementsValue
    ) as ElementStore;
    return {
      elements: getElements(rootElement, elementsValue),
      variables: [],
      config: {
        rootElementId,
        viewportWidth: configValue.viewportWidth,
        fonts: [],
        style: {},
      },
      relations: { elementFont: {} },
    };
  };
