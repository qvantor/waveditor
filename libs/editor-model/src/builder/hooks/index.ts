import { useContext, useMemo } from 'react';
import { BuilderContext } from '../../types';
import { BuilderContextValue } from '../constants';

export const useBuilderContext = (): BuilderContext => {
  const builder = useContext(BuilderContextValue);
  if (!builder)
    throw new Error('useBuilderContext used outside of BuilderProvider');
  return builder;
};

export const useAction = <R>(
  action: (context: BuilderContext) => R,
  deps: unknown[] = []
) => {
  const context = useBuilderContext();
  return useMemo(() => action(context), [context, ...deps]);
};
