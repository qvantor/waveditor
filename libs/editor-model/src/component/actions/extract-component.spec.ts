import { createBuilderContext } from '../../builder';
import { BasicCase } from '../../_tests';
import { extractComponent } from './extract-component';

const {
  Template,
  MooliFont,
  InterFont,
  AgeVariable,
  NameVariable,
  SimpleLayoutText,
  SimpleLayoutImage,
  ImageId,
  SimpleTextId,
  TextWithMooliFont,
  TextWithNameVariable,
  TextWithNameVariableAndMooliFont,
  TextWithTwoVariablesAndInterFont,
  SimpleLayoutOfThree,
  LayoutWith2FontsAnd2Variables,
} = BasicCase;

const builder = createBuilderContext(Template);
describe('extract component', () => {
  it('single image', () => {
    const component = extractComponent(builder)(ImageId);

    expect(Object.keys(component.elements).length).toBe(1);
    expect(component.elements[ImageId]).toEqual(Template.elements[ImageId]);
    expect(component.relations.elementFont).toEqual({});
    expect(component.variables.length).toBe(0);
    expect(component.config.fonts.length).toBe(0);
    expect(component.config.rootElementId).toBe(ImageId);
  });
  it('simple text', () => {
    const component = extractComponent(builder)(SimpleTextId);

    expect(Object.keys(component.elements).length).toBe(1);
    expect(component.elements[SimpleTextId]).toEqual(
      Template.elements[SimpleTextId]
    );
    expect(component.relations.elementFont).toEqual({});
    expect(component.variables.length).toBe(0);
    expect(component.config.fonts.length).toBe(0);
    expect(component.config.rootElementId).toBe(SimpleTextId);
  });

  it('text with font', () => {
    const component = extractComponent(builder)(TextWithMooliFont);

    expect(component.config.fonts).toEqual([MooliFont]);
    expect(component.variables.length).toBe(0);
    expect(component.relations.elementFont).toEqual({
      [TextWithMooliFont]: MooliFont.id,
    });
  });
  it('text with single variable', () => {
    const component = extractComponent(builder)(TextWithNameVariable);

    expect(component.config.fonts.length).toBe(0);
    expect(component.variables).toEqual([NameVariable]);
  });
  it('text with font and single variable', () => {
    const component = extractComponent(builder)(
      TextWithNameVariableAndMooliFont
    );

    expect(component.config.fonts).toEqual([MooliFont]);
    expect(component.variables).toEqual([NameVariable]);
  });
  it('text with font and couple variables', () => {
    const component = extractComponent(builder)(
      TextWithTwoVariablesAndInterFont
    );

    expect(component.elements[TextWithTwoVariablesAndInterFont]).toEqual(
      Template.elements[TextWithTwoVariablesAndInterFont]
    );
    expect(component.config.fonts).toEqual([InterFont]);
    expect(component.relations.elementFont).toEqual({
      [TextWithTwoVariablesAndInterFont]: InterFont.id,
    });
    expect(component.variables).toEqual([NameVariable, AgeVariable]);
  });
  it('simple layout of 3 elements', () => {
    const component = extractComponent(builder)(SimpleLayoutOfThree);

    expect(component.elements[SimpleLayoutOfThree]).toEqual(
      Template.elements[SimpleLayoutOfThree]
    );
    expect(component.elements[SimpleLayoutImage.id]).toEqual(SimpleLayoutImage);
    expect(component.elements[SimpleLayoutText.id]).toEqual(SimpleLayoutText);
    expect(Object.keys(component.elements).length).toBe(3);
    expect(component.config.rootElementId).toBe(SimpleLayoutOfThree);
    expect(component.config.fonts.length).toBe(0);
    expect(component.variables.length).toBe(0);
    expect(component.relations.elementFont).toEqual({});
  });
  it('layout with 2 font and 2 variables', () => {
    const component = extractComponent(builder)(LayoutWith2FontsAnd2Variables);
    expect(component.elements[LayoutWith2FontsAnd2Variables]).toEqual(
      Template.elements[LayoutWith2FontsAnd2Variables]
    );
    expect(Object.keys(component.elements).length).toBe(6);
    expect(component.config.fonts).toEqual([MooliFont, InterFont]);
    expect(component.variables).toEqual([NameVariable, AgeVariable]);
    expect(component.relations).toMatchSnapshot();
  });
});
