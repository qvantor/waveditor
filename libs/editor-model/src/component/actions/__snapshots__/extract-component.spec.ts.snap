// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`extract component layout with 2 font and 2 variables 1`] = `
Object {
  "elementFont": Object {
    "0.09364994320533238": "0.7248257932515505",
    "0.7402764635803678": "0.1332247839013243",
  },
}
`;
